ukvm/ukvm_module_dumpcore.c

```diff
@@ -0,0 +1,218 @@
+ * vmcore.
+ */
+void ukvm_dumpcore(struct ukvm_hv *hv, struct ukvm_halt *info)
+{
+    int core_fd, num_notes = 0;
+    size_t offset, note_size;
+    Elf64_Ehdr hdr;
+    Elf64_Phdr phdr;
+    char filename[20] = {0};
+    snprintf(filename, 20, "core.ukvm.%d", getpid());
+
+    memset((void *)&hdr, 0, sizeof(Elf64_Ehdr));
+    memset((void *)&phdr, 0, sizeof(Elf64_Phdr));
+
+    core_fd = open(filename, O_RDWR|O_CREAT|O_TRUNC|O_APPEND, S_IRUSR|S_IWUSR);
+    if (core_fd < 0) {
+        goto failure;
+    }
+
+    memset(&hdr, 0, sizeof(Elf64_Ehdr));
+    memcpy(&hdr, ELFMAG, SELFMAG);
+
+    hdr.e_ident[EI_VERSION] = EV_CURRENT;
+    hdr.e_type = ET_CORE;
+    hdr.e_version = EV_CURRENT;
+    hdr.e_machine = EM_X86_64;
+    hdr.e_ehsize = sizeof(hdr);
+    hdr.e_phoff = sizeof(hdr); 
+    hdr.e_phentsize = sizeof(Elf64_Phdr);
+
+    /* Fill in architecture specific contents for elf header */
+    ukvm_dumpcore_fill_arch_header(&hdr);
+
+    /* No section header in core file */
+    hdr.e_shoff = 0;
+    hdr.e_shentsize = 0;
+    hdr.e_shnum = 0;
+    hdr.e_shstrndx = 0;
+
+    note_size = ukvm_dumpcore_get_note_size(&num_notes);
+    hdr.e_phnum = 1 + num_notes;
+
+    if (write(core_fd, &hdr, sizeof(Elf64_Ehdr)) < 0) {
+        goto failure;
+    }
+
+    offset = sizeof(hdr) + (hdr.e_phnum * sizeof(Elf64_Phdr));
+
+    if (num_notes) {
+        /* Write a core PT_NOTE, data for more useful notes is not available for now */
+        phdr.p_type = PT_NOTE;
+        phdr.p_offset = offset;
+        phdr.p_filesz = note_size;
+        phdr.p_memsz = note_size;
+        if (write(core_fd, &phdr, sizeof(phdr)) < 0 ) {
+            goto failure;
+        }
+        offset += note_size;
+    }
+
+    /* Write program headers for memory segments. We have only one? */
+    memset(&phdr, 0, sizeof(phdr));
+    phdr.p_type     = PT_LOAD;
+    phdr.p_align    = 0;
+    phdr.p_paddr    = (size_t)hv->mem;
+    phdr.p_offset = offset;
+    phdr.p_vaddr  = 0;
+    phdr.p_memsz  = hv->mem_size;
+    phdr.p_filesz = hv->mem_size;
+    phdr.p_flags  = 0;
+
+    if (write(core_fd, &phdr, sizeof(phdr)) < 0) {
+        goto failure;
+    }
+
+    /* Write note section */
+    if (num_notes &&
+        ukvm_dumpcore_dump_notes(core_fd, hv, info) < 0) {
+        goto failure;
+    }
+
+    /* Write memory */
+    if (write(core_fd, hv->mem, hv->mem_size) < 0) {
+        goto failure;
+    }
+    warnx("Dumped an unikernel core file: %s", filename);
+    goto cleanup;
+
+failure:
+    warnx("%s: Failed to generate the core file", __FUNCTION__);
+
+cleanup:
+    close(core_fd);
+}
+
+static void hypercall_dumpcore(struct ukvm_hv *hv, ukvm_gpa_t gpa)
+{
+    if (use_dumpcore) {
+        struct ukvm_halt *t =
+            UKVM_CHECKED_GPA_P(hv, gpa, sizeof (struct ukvm_halt));
+
+        if (ukvm_dumpcore_get_regs(hv) == 0) {
+            ukvm_dumpcore(hv, t);
+        }
+    } else {
+        warnx("Not dumping core since --dumpcore option is not set");
```

Comment:
Better not to print anything. This module will be enabled on most cases, and this will get printed even if the unikernel exits cleanly (rc=0).